Plans 2023: Always show the biannual toggle. free/paid site plans

chriskmnds opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes https://github.com/Automattic/wp-calypso/issues/74636
Based on https://github.com/Automattic/wp-calypso/pull/74722

## Proposed Changes

This will show the 1yr/2yr toggle on both paid/free site plans (irrespective of billing term) when visiting `/plans/[SITE]` with the `plans/biannual-toggle` flag enabled. 

Previously, the toggle was only visible on free sites and on `/plans/monthly/[SITE]` with site on a monthly plan due to the "monthly" option and monthly restrictions (downgrading from yearly to monthly not being permissible).

### Media

<img width="700" alt="Screenshot 2023-03-22 at 4 04 28 PM" src="https://user-images.githubusercontent.com/1705499/226931799-fda7f70f-57c3-4403-9310-62c70928b280.png">


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to `plans/[SITE]` with both free/paid sites and feature flag enabled `?flags=plans/biannual-toggle`
* The 1yr/2yr toggle should always be visible/active

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

chriskmnds committed:
add passlist for interval types in plans controller

chriskmnds committed:
cleanup. include triannual plans in comparison grid sort

chriskmnds committed:
Plans 2023: Always show biannual toggle. free/paid site plans

aneeshd16 reviewed (commented):
I tested this with a site on a monthly plan and noticed that while the toggle was visible, none of the toggle options were in the active state. Also, the plans grid showed monthly prices. 
<img width="1918" alt="image" src="https://user-images.githubusercontent.com/5436027/227126912-d3143156-c555-4295-9947-8a45db99709e.png">

I can't think of any alternate solutions for this other than having a 3-way toggle with Pay monthly/Pay 1 year/Pay 2 years. 🤔

chriskmnds committed:
redirect /plans -> /plans/yearly

chriskmnds commented:
> I tested this with a site on a monthly plan and noticed that while the toggle was visible, none of the toggle options were in the active state. Also, the plans grid showed monthly prices.

Good catch @aneeshd16.  This happens only on `/plans/[SITE]` with a site on monthly, right?

> I can't think of any alternative solutions for this other than having a 3-way toggle with Pay monthly/Pay 1 year/Pay 2 years. 🤔

Do you recon there would be any problems with redirecting to `/plans/yearly` in that case? You can test this behavior from the latest commit.

aneeshd16 commented:
> This happens only on /plans/[SITE] with a site on monthly, right?

That's right. 👍 

> Do you recon there would be any problems with redirecting to /plans/yearly in that case? You can test this behavior from the latest commit.

I tried with the latest commit:
1. The "Manage Plan" CTA and the "Your plan" badge are shown when I switch to the 2-yearly view (the site is on a monthly Premium plan)
2. The "Personal" plan CTA should be an "Upgrade" CTA instead of the "Downgrade" one.
<img width="1920" alt="image" src="https://user-images.githubusercontent.com/5436027/227505988-2557a34e-dc82-4eaa-a208-f040f5f67589.png">

Even if we fix the above two points, this will lead to a product change:
1. Users won't see the "Manage Plan" CTA and the "Your plan" badge.
2. Users won't be able to upgrade to a higher monthly plan.

chriskmnds commented:
> The "Manage Plan" CTA and the "Your plan" badge are shown when I switch to the 2-yearly view (the site is on a monthly Premium plan)

This should be fixed in https://github.com/Automattic/wp-calypso/pull/74722. From my end when visiting `/plans/2yearly/[MONTHLY_PREMIUM]` (on https://github.com/Automattic/wp-calypso/pull/74722):

<img width="700" alt="Screenshot 2023-03-24 at 2 01 35 PM" src="https://user-images.githubusercontent.com/1705499/227516199-05e2202b-15d9-4961-8a78-ea5c96de8fac.png">

> The "Personal" plan CTA should be an "Upgrade" CTA instead of the "Downgrade" one.

This is also the case in trunk right now when visiting `/plans/yearly/[MONTHLY]`. It's not something that is being addressed anywhere AFAIK. Also not sure what/ if any caveats with "upgrading" to a lower tier plan for longer-term 🤔  From my end with monthly premium, on trunk:

<img width="700" alt="Screenshot 2023-03-24 at 2 07 47 PM" src="https://user-images.githubusercontent.com/1705499/227517227-555962b1-bb46-4945-83f7-a8d4fbe7194b.png">

> Users won't see the "Manage Plan" CTA and the "Your plan" badge.

I think we settled on not treating longer-term plans as "current plan" in the grids for shorter-term ones. Same for looking at 2-yearly with a yearly plan, etc. This is what is covered by in https://github.com/Automattic/wp-calypso/pull/74722 (and the reason it exists)

> Users won't be able to upgrade to a higher monthly plan.

Indeed. I think this is the main deviation (and potential blocker) with having the redirection (`/plans/monthly` is still accessible/usable, but not linked from the toggle). I'll get back to this and see if any workaround.

chriskmnds committed:
unit/ui tests for upgrade actions

chriskmnds committed:
unit/ui tests for upgrade actions

chriskmnds committed:
clean up mocks. move variable mock implementation to beforeEach cb. fails ts check but tests pass- needs more investigation

chriskmnds committed:
Revert "redirect /plans -> /plans/yearly"

This reverts commit 3dd3ad05b84a7b787a6d332773e5caa32335911e.

southp reviewed (changes requested):
I've confirmed that:

* When the 1y/2y toggle is on, the behavior of /plans for the monthly plan owners match with what we've agreed on: the toggle looks untoggled to either side, and the monthly plan grid shows by default.
* When the 1y/2y toggle is off, /plans works the way it is.
* For an annual plan owner account, everything works either the 1y/2y toggle is on or off.

However, regarding this:
> Also not sure what/ if any caveats with "upgrading" to a lower tier plan for longer-term 
It is a forbidden behavior from the backend. Thus, when I attempt to "upgrade" from a 2y Premium plan to a 1y Business plan, the following error will show:
<img width="1189" alt="截圖 2023-03-31 下午4 36 53" src="https://user-images.githubusercontent.com/1842898/229072237-fdf3cd42-dfef-44d4-b7f9-14d3707a5894.png">

The simplest solution I can think of now is to, once again, hide the toggle when a user is on a 2y plan. I'm wondering if we'd need to generalize it a bit, e.g. if the highest term the toggle poses is equal or lower to the user's plan, don't show the toggle. I'll leave that to your best judgement, of course; just thinking out loud here :)

chriskmnds commented:
Thanks, @southp.

> Also not sure what/ if any caveats with "upgrading" to a lower tier plan for longer-term
> It is a forbidden behavior from the backend. Thus, when I attempt to "upgrade" from a 2y Premium plan to a 1y Business plan, the following error will show:

(Indeed i thought about it but I guess my brain too limited to connect it to the 2y part 😆 )

> The simplest solution I can think of now is to, once again, hide the toggle when a user is on a 2y plan. I'm wondering if we'd need to generalize it a bit, e.g. if the highest term the toggle poses is equal or lower to the user's plan, don't show the toggle. I'll leave that to your best judgement, of course; just thinking out loud here :)

This might be the simplest way out indeed. However, also thinking about whether it makes sense to instead give users more visibility and a unified experience with this e.g. show the toggle to all, but change the active states of buttons. so 1y, 2y, 3y users will see the same, just not be able to proceed with the "forbidden" resource.

chriskmnds committed:
contact support for lower-term higher-tier plans

chriskmnds committed:
cleanup and tests

chriskmnds committed:
getPlanBillPeriod tests

chriskmnds committed:
do not show toggle on sites with > 1 year plan

chriskmnds commented on client/my-sites/plans-features-main/plan-type-selector.tsx:
```diff
@@ -199,26 +213,6 @@ export const IntervalTypeToggle: React.FunctionComponent< IntervalTypeProps > =
 	);
 };
 
-export const ExperimentalIntervalTypeToggle: React.FunctionComponent< IntervalTypeProps > = (
-	props
-) => {
-	return (
-		<ProvideExperimentData name="calypso_show_interval_type_selector_2022_07">
-			{ ( isLoading, experimentAssignment ) => {
-				if ( isLoading || ! props.intervalType ) {
-					return <></>;
-				}
-
-				if ( 'treatment' !== experimentAssignment?.variationName ) {
-					return <></>;
-				}
-
-				return <IntervalTypeToggle { ...props } />;
-			} }
-		</ProvideExperimentData>
-	);
-};
-
```
robots 😆

chriskmnds committed:
cleanup. bring back deleted experiment code

chriskmnds committed:
cleanup. fix failing test

chriskmnds commented:
Thanks again for reviewing/testing. 😎 

> The simplest solution I can think of now is to, once again, hide the toggle when a user is on a 2y plan. I'm wondering if we'd need to generalize it a bit, e.g. if the highest term the toggle poses is equal or lower to the user's plan, don't show the toggle. I'll leave that to your best judgement, of course; just thinking out loud here :)

This should hopefully be handled now. Updated to do two things:

- Not show the biannual toggle when site has a plan > 1 year.
- Show a disabled "Contact support" button (instead of an active "Upgrade" button) when site is on a lower tier but higher term plan e.g. current site plan 2Y Premium and grid plan 1Y Business - Only relevant when visiting the paths directly e.g. `/plans/yearly/[SITE_2Y_PLAN]`.

Rebased onto https://github.com/Automattic/wp-calypso/pull/74722 to make use of the action button changes.

The proposed changes are also temporary until we have a decision on the toggle. The "showBiannualToggle" prop is restrictive, but refactoring now will mean touching code with the running experiment and redoing the new experiment code. I will create a follow-up issue to come back to this and make things more flexible to incorporate 3Y term in the toggle.

cc @Automattic/martech-decepticons

southp reviewed (approved):
Thanks for the fix, @chriskmnds . I've re-confirmed all the combinations and they all work as expected. Please feel free to ship this after resolving the code conflicts :)

chriskmnds committed:
cleannup tests
